### ISSUE TITLE
label graphs via prairiedog as req by diffpool

### ISSUE THREAD

**Issue by kevinkle** [superphy/prairiedog#52]
namely:

- G.graph['label']
- G.node[u]['label']
- G.node[u]['feat']
- G.graph['feat_dim']

and

- nodes need to be indexed from 0
- node labels need to be one-hot encoded

**Comment by kevinkle** [comment]
looks like node ids dont have to be unique between graphs (ie any 2 graphs can start with node id 0) and we'll just work off the kmer id mapping for labels

**Comment by kevinkle** [comment]
Note that node "label" should actually be labelled "feat" anyways
```
    if feat == 'node-feat' and 'feat_dim' in graphs[0].graph:
        print('Using node features')
        input_dim = graphs[0].graph['feat_dim']
    elif feat == 'node-label' and 'label' in graphs[0].node[0]:
        print('Using node labels')
        for G in graphs:
            for u in G.nodes():
                G.node[u]['feat'] = np.array(G.node[u]['label'])
```

**Comment by kevinkle** [comment]
Node labels are also indexed from 0
```
node_labels=[]
    try:
        with open(filename_nodes) as f:
            for line in f:
                line=line.strip("\n")
                node_labels+=[int(line) - 1]
```

**Comment by kevinkle** [comment]
Graph labels look to be indexed from 1

**Comment by kevinkle** [comment]
One hot encoding of kmer labels is causing us memory problems, likely because diffpool is doing it w a numpy array. Thinking of trying https://discuss.pytorch.org/t/convert-int-into-one-hot-format/507/33

**Comment by kevinkle** [comment]
Looks like the correct approach is probably to use a nn.Embedding https://discuss.pytorch.org/t/best-way-to-train-on-one-hot-vectors/29127/5

**Comment by kevinkle** [comment]
From ours:
```
  m.bias.data = init.constant(m.bias.data, 0.0)
/home/kevin/diffpool/encoders.py:293: UserWarning: nn.init.xavier_uniform is now deprecated in favor of nn.init.xavier_uniform_.
  m.weight.data = init.xavier_uniform(m.weight.data, gain=nn.init.calculate_gain('relu'))
/home/kevin/diffpool/encoders.py:295: UserWarning: nn.init.constant is now deprecated in favor of nn.init.constant_.
  m.bias.data = init.constant(m.bias.data, 0.0)
Epoch:  0
Creating a zero array with max_num_nodes 19189 and feat_dim 1
Creating a zero array with max_num_nodes 19189 and feat_dim 1
Creating a zero array with max_num_nodes 19189 and feat_dim 1
Seeing data['label'].long() as tensor([1, 2, 2])
/home/kevin/prairiedog/venv/lib/python3.6/site-packages/torch/nn/_reduction.py:46: UserWarning: size_average and reduce args will be deprecated, please use reduction='mean' instead.
  warnings.warn(warning.format(ret))
/home/kevin/diffpool/train.py:221: UserWarning: torch.nn.utils.clip_grad_norm is now deprecated in favor of torch.nn.utils.clip_grad_norm_.
```

**Comment by kevinkle** [comment]
original:
```
Seeing data['label'].long() as tensor([1, 5, 1])
Seeing data['feats'].float() as tensor([[[0., 1., 0.],
         [1., 0., 0.],
         [0., 1., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[1., 0., 0.],
         [1., 0., 0.],
         [1., 0., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]],

        [[0., 1., 0.],
         [0., 1., 0.],
         [0., 1., 0.],
         ...,
         [0., 0., 0.],
         [0., 0., 0.],
         [0., 0., 0.]]])
```

**Comment by kevinkle** [comment]
Ours:
```
Seeing data['feats'].float() as tensor([[[3242491.],
         [ 387054.],
         [1998565.],
         ...,
         [2859968.],
         [2014661.],
         [2184725.]],

        [[2604603.],
         [2029807.],
         [1332638.],
         ...,
         [      0.],
         [      0.],
         [      0.]],

        [[1829966.],
         [3125562.],
         [1998535.],
         ...,
         [      0.],
         [      0.],
         [      0.]]])
Seeing data['label'].long() as tensor([ 1,  2, 11])
/home/kevin/prairiedog/venv/lib/python3.6/site-packages/torch/nn/_reduction.py:46: UserWarning: size_average and reduce args will be deprecated, please use reduction='mean' instead.
  warnings.warn(warning.format(ret))
THCudaCheck FAIL file=/pytorch/aten/src/THCUNN/generic/ClassNLLCriterion.cu line=224 error=59 : device-side assert triggered
/pytorch/aten/src/THCUNN/ClassNLLCriterion.cu:105: void cunn_ClassNLLCriterion_updateOutput_kernel(Dtype *, Dtype *, Dtype *, long *, Dtype *, int, int, int, int, long) [with Dtype = float, Acctype = float]: block: [0,0,0], thread: [2,0,0] Assertion `t >= 0 && t < n_classes` failed.
```
Its prob the lack of one-hot encoding for kmer labels

**Comment by kevinkle** [comment]
Ours:
```
Epoch:  0
Creating a zero array with max_num_nodes 19977 and feat_dim 1
Going to return 'feature' as [[3242491.]
 [ 387054.]
 [1998565.]
 ...
 [2859968.]
 [2014661.]
 [2184725.]]
Creating a zero array with max_num_nodes 19977 and feat_dim 1
Going to return 'feature' as [[1398149.]
 [ 190656.]
 [3050501.]
 ...
 [      0.]
 [      0.]
 [      0.]]
Creating a zero array with max_num_nodes 19977 and feat_dim 1
Going to return 'feature' as [[ 227776.]
 [ 186124.]
 [1109563.]
 ...
 [      0.]
 [      0.]
 [      0.]]
Seeing data['feats'].float() as tensor([[[3242491.],
         [ 387054.],
         [1998565.],
         ...,
         [2859968.],
         [2014661.],
         [2184725.]],

        [[1398149.],
         [ 190656.],
         [3050501.],
         ...,
         [      0.],
         [      0.],
         [      0.]],

        [[ 227776.],
         [ 186124.],
         [1109563.],
         ...,
         [      0.],
         [      0.],
         [      0.]]])
Seeing data['label'].long() as tensor([1, 2, 2])
/home/kevin/prairiedog/venv/lib/python3.6/site-packages/torch/nn/_reduction.py:46: UserWarning: size_average and reduce args will be deprecated, please use reduction='mean' instead.
  warnings.warn(warning.format(ret))
/home/kevin/diffpool/train.py:222: UserWarning: torch.nn.utils.clip_grad_norm is now deprecated in favor of torch.nn.utils.clip_grad_norm_.
  nn.utils.clip_grad_norm(model.parameters(), args.clip)
```

**Comment by kevinkle** [comment]
```
(venv) [kevinle@waffles prairiedog]$ sbatch -c 6 --mem 200G --partition NMLResearch --wrap="snakemake -j 1"
Submitted batch job 6888945
```

**Comment by kevinkle** [comment]
https://github.com/dmlc/dgl/issues/459
https://github.com/dmlc/dgl/pull/439

**Comment by kevinkle** [comment]
`wontfix` per https://github.com/superphy/prairiedog/issues/48#issuecomment-503740468
